File: BookShelf/Services/BookService.cs
using BookShelf.Model;
using System.Collections.Generic;

namespace BookShelf.Services
{

    // Interface for BookService operations
    public interface BookService
    {
        //Retrieves a list of all books in the collection
        List<Book> GetAllBooks();

        //Retrieves a specific book by its unique identifier (i
        Book GetBookById(int id);

        //Adds a new book to the collection and returns the added book
        Book AddBook(Book book);

        // Updates an existing book based on its id and returns the updated book
        Book UpdateBook(int id, Book book);

        // Deletes a book from the collection based on its id
        bool DeleteBook(int id);
    }
}

File: BookShelf-FE/book-shelf/src/app/services/book.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Book } from '../models/book.model';

@Injectable({
  providedIn: 'root'
})
export class BookService {

  //URL for the backend service
  private apiUrl = 'https://localhost:44327/api/Books';

// Injecting HttpClient service
  constructor(private http: HttpClient) { }

  //fetches all books from the backend
  getBooks(): Observable<Book[]> {
    return this.http.get<Book[]>(this.apiUrl);
  }

  //Fetches a single book by its ID
  getBook(id: number): Observable<Book> {
    return this.http.get<Book>(`${this.apiUrl}/${id}`);
  }

  //Adds a new book
  addBook(book: Book): Observable<Book> {
    return this.http.post<Book>(this.apiUrl, book);
  }

  //Updates an existing book's details
  updateBook(book: Book): Observable<Book> {
    return this.http.put<Book>(`${this.apiUrl}/${book.id}`, book);
  }

  //Deletes a book from the backend
  deleteBook(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
